File: 01-week-1--starter-algorithms/03-day-4--selection-sort/javascript/selection_sort.js
function selectionSort(arr) {
  // type your code here
  /*
  if (arr.length == 0 || arr.length == 1) {
    return arr
  } else {
    return arr.sort(function(a, b) {
      return a - b
    })
  }
  */

  const populateKeyValues = () => {
    let i = 0
    const obj = {}
    arr.forEach((element) => {
      obj[element] = i
      if (i == arr.length - 1) {
        return sortKeyValues(obj)
      } else {
        i++
      }
    })
  }

  const sortKeyValues = (obj) => {
    let i = 1
    arr.forEach((element) => {
      const next = (!!arr[i] ? arr[i] : false )
      if (next) {
        if (element < next && obj[element] > obj[next]) {
            let lesserIndex = obj[next]
            let greaterIndex = obj[element]
            obj[element] = lesserIndex
            obj[next] = greaterIndex
            i++
        }
      } else {
        return objToIndexObj(obj)
      }
    })
  }

  const objToIndexObj = (obj) => {
    let indexObj = {}
    arr.forEach(value => {
      indexObj[obj[value]] = value
      if (obj[value] == arr.length - 1) {
        return indexObjToArray(indexObj)
      }
    })
  }

  const indexObjToArray = (indexObj) => {
    let answerArr = []
    for (let i = 0; i == arr.length; i++) {
      answerArr.push(indexObj[i])
      if (i == arr.length - 1) {
        return answerArr
      }
    }
  }

  return populateKeyValues()
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: [-1, 2, 3, 5]");
  console.log("=>", selectionSort([3, -1, 5, 2]));

  console.log("");

  // BENCHMARK HERE, and print the average runtime
  const longInput = [];

  for (let i = 0; i < 100; ++i) {
    longInput.push(Math.random());
  }
}

module.exports = selectionSort;

// Please add your pseudocode to this file
/*
  -create a blank array
  -fill this array with elements representing indexes that is equal in length to the argument array
  -create a blank object
  -create a variable, set it to zero
  -iterate through the argument array, assign each key using elements from the created array and the values using
   the argument array
  -create another blank array
  -create a function that finds the lowest value in the hash if the value is true, pushes that number to the array, and 
   sets that hash keys value to false.
  -return this new array
*/
// And a written explanation of your solution

File: 01-week-1--starter-algorithms/03-day-4--selection-sort/ruby/selection_sort.rb
def selection_sort(arr)
  # type your code in here
  if arr.length == 0 || arr.length == 1
    arr
  else
    arr_copy = arr
    answer = []
    while arr_copy.length > 0 do
      min = arr_copy.min
      i = arr_copy.index(min)
      arr_copy.delete_at(i)
      answer.push(min)
    end
  answer
  end
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: [-1, 2, 3, 5]"
  print "=> "
  print selection_sort([3, -1, 5, 2])

  puts

  # Don't forget to add your own!

  # BENCHMARK HERE, you can print the average runtime
  long_input = []

  100.times { long_input << rand }
end

# Please add your pseudocode to this file
=begin
  -if arr.length equals 0 or 1 return arr
  -else, create copy of arr *arr_copy, create empty array *answer
  -while arr_copy's length is more than 0 push arr_copy's minimum element to answer,
   then delete the element at that index in arr_copy
  -return answer
=end
# And a written explanation of your solution
=begin
  I wanted to create a copy of the argument and push each minimum value to an answer array while destroying the element at that index
  within the copy of the argument array.  At the end I wanted to return the answer array.
=end
